apollo-router/src/query_planner/dual_query_planner.rs

```diff
@@ -726,28 +733,45 @@ fn same_ast_document(x: &ast::Document, y: &ast::Document) -> Result<(), MatchFa
         "Different number of operation definitions"
     );
 
+    check_match_eq!(x_frags.len(), y_frags.len());
+    let mut fragment_map: HashMap<Name, Name> = HashMap::new();
+    // Assumption: x_frags and y_frags are topologically sorted.
+    //             Thus, we can build the fragment name mapping in a single pass and compare
+    //             fragment definitions using the mapping at the same time, since earlier fragments
+    //             will never reference later fragments.
+    x_frags.iter().try_fold((), |_, x_frag| {
```

Comment:
Not important, but I think you're looking for `.try_for_each` rather than `.try_fold` using a unit.

Reply:
Well, I wanted to something like `try_all` (https://github.com/rust-lang/rfcs/pull/3233) or `FallibleIterator` (but it's private). We should make `FallibleIterator` more accessible :)